### PR TITLE
Create plugin support for CDN'ing of static assets.

### PR DESCRIPTION
**Pull request by georgestephanis**
Add in CDN hosted versions of Core / Jetpack scripts and styles to speed up sites and better cross-site cacheability.

#### To do:

- [x] Add caching to external http calls so they're not requested on each load.
- [x] Abstract out plugin script/style swapping to enable for other plugins (like WooCommerce) as well.
- [x] Confirm RTL styles are handled correctly.

#### Testing instructions:

* Activate Photon CDN through the old module settings. Confirm that scripts / styles are now pointing to c0.wp.com resources in the page source. If testing using a development version, you can use a filter to override versions to production versions:
```
add_filter( 'jetpack_cdn_plugin_slug_and_version', function( $values ) {
	return array( 'jetpack', '6.5' );
} );
add_filter( 'jetpack_cdn_core_version_and_locale', function( $values ) {
	return array( '4.9.7', 'en_US' );
} );
```
* Make sure everything loads in a couple browsers.
* Test in multisite
* Test in subfolder installs.
* Test in RTL

#### Proposed changelog entry for your changes:

* Speed up sites and increase max concurrent connections through Photon by cloud-hosting Jetpack and WordPress Core scripts, styles, and assets in addition to content images.

### CODE REVIEW

**Comment by georgestephanis** [comment]
I'm unsure if it may be wiser to change how we swap the urls and such to a just in time based off a later filter or action.

**Comment by georgestephanis** [comment]
@matticbot Go away. 😜

**Comment by georgestephanis** [comment]
@zinigor There seems to have been some unintended whitespace changes from when you regenerated `modules/module-headings.php` ?  My guess is it was a change in line endings?  Still, probably needless churn if it can be fixed.

**Comment by georgestephanis** [comment]
After chatting with @bazza I'm going to take a swing at writing a manifest builder script for Jetpack that will generate a json manifest of js/css files in the plugin that we can swap out to the cdn variant, rather than hitting .org

I'm also going to try to see if we can just swap any css/js file in wp-admin or wp-includes to the cdn instead -- I'm not sure if that's the ideal solution, but it may speed things up and result in less stuff stored in the options table.

File: modules/photon-cdn.php [hunk 1]
```diff
@@ -108,6 +123,19 @@ public static function get_core_assets( $version = null, $locale = null ) {
...
+		/**
+		 * Filters Jetpack CDN's Core version number and locale. Can be used to override the values
+		 * that Jetpack uses to retrieve assets. Expects the values to be returned in an array.
+		 *
+		 * @since 6.6
+		 *
+		 * @param array $values array( $version  = core assets version, i.e. 4.9.1, $locale = desired locale )
+		 */
+		list( $version, $locale ) = apply_filters(
+			'jetpack_cdn_plugin_slug_and_version',
```
Comment (georgestephanis) [inline modules/photon-cdn.php]: this one runs different parameters than the other, and should probably have a different filter name.  `jetpack_cdn_core_version_and_locale` maybe?

Comment (zinigor) [reply modules/photon-cdn.php]: Yes, copypasta, good catch, thank you!

**Comment by jeherve** [comment]
How do we know if a file should be served by the CDN? I have, for example, quite a few files that don't seem to be served by the CDN in wp-admin, but I don't know if that's on purpose?

screenshot 2018-09-17 at 13 28 23

**Comment by georgestephanis** [comment]
Aha, I think that's because I was running off the front-end hooks.  Lemme take a swing at adding admin side accelleration too.

**Comment by georgestephanis** [comment]
Yarp.  https://github.com/Automattic/jetpack/pull/10102/commits/1f6e7dc9622a7abce850d39d2f3620857b13acb1 did it.

**Comment by georgestephanis** [comment]
RTL seems to be working a-ok.

**Comment by jeherve** [comment]
Things look really good.

I can see that `wp-emoji-release.min.js` (or `wp-emoji.js` and `twemoji.js` if you have `SCRIPT_DEBUG` set) is still served locally, but everything else from Jetpack and Core appears to be served by the CDN now. The increase in speed is definitely noticeable!

I noticed what seems to be a conflict with the Akismet plugin. When in `wp-admin`, we try to load `https://c0.wp.com/c/4.9.7/wp-content/plugins/akismet/_inc/akismet.css` and `https://c0.wp.com/c/4.9.7/wp-content/plugins/akismet/_inc/akismet.js`, but those files do not exist.

**Comment by kraftbj** [comment]
The emoji files are loaded differently than most https://github.com/WordPress/WordPress/blob/56c162fbc9867f923862f64f1b4570d885f1ff03/wp-includes/formatting.php#L5431

We'd need to use the script_loader_src filter for that: https://developer.wordpress.org/reference/hooks/script_loader_src/ targeting the `wpemoji`, `twemoji`, `concatemoji` handles.

**Comment by georgestephanis** [comment]
I'm probably okay skipping exceptions for the emoji ones for the moment -- I don't want the initial PR to be too epically complex.

Good catch on Akismet -- I think that's because it's coming back with the core hashes, as it's bundled in the zip -- but not getting pulled over by @bazza's cdn.  Not a big deal, I'll just prune them from what gets rewritten.

Talking to @claudiulodro about WooCommerce in Slack currently, seeing when we can get WC core onboard.  They're currently past feature freeze, so it'll need to probably be for 3.6

**Comment by zinigor** [comment]
@georgestephanis sorry, there seems to be a problem with your latest change. It looks like Jetpack, and maybe some other plugins as well, enqueue assets with URLs like this:
umz2zkxmqy-3000x3000

We're going to be breaking these for all plugins that rely on that mechanism. We could probably blacklist URLs that have `admin-ajax` in them, but that sounds risky because other plugins might do something similar with a different URL. What's your opinion on this?

**Comment by georgestephanis** [comment]
Ahhh shucks.  Okay.  Good find, I'll add a check.

**Comment by georgestephanis** [comment]
That should have it fixed, @zinigor

**Comment by zinigor** [comment]
Tested, fixes the problem, thank you! And huge props for @dereksmart for finding the problem!

**Comment by jeherve** [comment]
I tested the PR one more time, and did not run into any major issues. It works well.

One quick note is that you can't really test this when using the Lazy Images feature; since we've modified Lazy Images in the past few weeks, the .js file served from the CDN is not compatible with the current development version of Jetpack.

**Comment by georgestephanis** [comment]
Once it's launchable, I'd like to switch it to only cdn on public versions of the codebase.  Should remedy that issue, but possibly make it harder to test.

**Pull request merged**